### PR TITLE
Enable multi-trajectory on ROS node

### PR DESCRIPTION
**Pull request by yutakaoka**
I added a start_trajectory service to create another trajectory. Also edited finish_trajectory. Please review it.

### CODE REVIEW

File: cartographer_ros/cartographer_ros/node_options.cc [hunk 1]
```diff
@@ -24,43 +24,13 @@ NodeOptions CreateNodeOptions(
...
 
-  CHECK_EQ(options.use_laser_scan + options.use_multi_echo_laser_scan +
-               (options.num_point_clouds > 0),
-           1)
-      << "Configuration error: 'use_laser_scan', "
-         "'use_multi_echo_laser_scan' and 'num_point_clouds' are "
-         "mutually exclusive, but one is required.";
-
-  if (options.map_builder_options.use_trajectory_builder_2d()) {
+  if (options.map_options.map_builder_options.use_trajectory_builder_2d()) {
```
Comment (SirVer) [inline cartographer_ros/cartographer_ros/node_options.cc]: This check must be done for every newly added trajectory and can be removed here. The whole function can then be inlined at the call sites and the NodeOptions struct be removed. MapOptions can then be renamed back to NodeOptions.

File: cartographer_ros/cartographer_ros/node.cc [hunk 2]
```diff
@@ -45,22 +47,25 @@ namespace carto = ::cartographer;
...
+constexpr double kTfBufferCacheTimeInSeconds = 1e6;
 constexpr int kLatestOnlyPublisherQueueSize = 1;
 
-Node::Node(const MapOptions& map_options, tf2_ros::Buffer* const tf_buffer)
-    : map_options_(map_options), map_builder_bridge_(map_options_, tf_buffer) {}
+TrajectoryOptions LoadTrajectoryOptions(const string& code) {
+  auto file_resolver = cartographer::common::make_unique<
+      cartographer::common::ConfigurationFileResolver>(
+      std::vector<string>{""});
+  cartographer::common::LuaParameterDictionary lua_parameter_dictionary(
```
Comment (SirVer) [inline cartographer_ros/cartographer_ros/node.cc]: After understanding the code a bit better, I think this should not be needed if `TrajectoryOptions` is converted to a ROS message with a binary proto.

Comment (yutakaoka) [inline cartographer_ros/cartographer_ros/node.cc]: Done

File: cartographer_ros/cartographer_ros/node.cc [hunk 3]
```diff
@@ -87,6 +92,33 @@ void Node::Initialize() {
...
       ::ros::WallDuration(map_options_.pose_publish_period_sec),
       &Node::PublishTrajectoryStates, this));
+
+  start_trajectory_server_ = node_handle_.advertiseService(
+      kStartTrajectoryServiceName, &Node::HandleStartTrajectory, this);
+  finish_trajectory_server_ = node_handle_.advertiseService(
+      kFinishTrajectoryServiceName,
+      &Node::HandleFinishTrajectory, this);
+  write_assets_server_ = node_handle_.advertiseService(
+      kWriteAssetsServiceName,
```
Comment (yutakaoka) [inline cartographer_ros/cartographer_ros/node.cc]: Done

File: cartographer_ros/cartographer_ros/node.cc [hunk 4]
```diff
@@ -185,4 +217,207 @@ void Node::SpinOccupancyGridThreadForever() {
...
+                    << request.trajectory_id << " is not created yet.");
+    return false;
+  }
+  if (!is_active_trajectory_[request.trajectory_id]) {
+    ROS_INFO_STREAM("Trajectory_id "
+                    << request.trajectory_id << " has already been finished.");
+    return false;
+  }
+
+  if (laser_scan_subscribers_.count(request.trajectory_id) != 0) {
```
Comment (yutakaoka) [inline cartographer_ros/cartographer_ros/node.cc]: Done

File: cartographer_ros/cartographer_ros/node.h [hunk 5]
```diff
@@ -64,7 +72,19 @@ class Node {
   bool HandleSubmapQuery(
       cartographer_ros_msgs::SubmapQuery::Request& request,
       cartographer_ros_msgs::SubmapQuery::Response& response);
-
+  bool HandleStartTrajectory(
```
Comment (yutakaoka) [inline cartographer_ros/cartographer_ros/node.h]: Done

File: cartographer_ros/cartographer_ros/node.h [hunk 6]
```diff
@@ -75,16 +95,23 @@ class Node {
...
-  std::unordered_set<string> expected_sensor_ids_;
 
   ::ros::NodeHandle node_handle_;
   ::ros::Publisher submap_list_publisher_;
   ::ros::ServiceServer submap_query_server_;
   ::ros::Publisher scan_matched_point_cloud_publisher_;
   cartographer::common::Time last_scan_matched_point_cloud_time_ =
       cartographer::common::Time::min();
-
+  ::ros::ServiceServer start_trajectory_server_;
```
Comment (yutakaoka) [inline cartographer_ros/cartographer_ros/node.h]: Yes, these need to live for the lifetime of the node just the same as other services(e.g. finish trajectory). Should I use a vector for all services?

File: cartographer_ros/cartographer_ros/trajectory_options.h [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+#include <string>
+
+#include "cartographer/common/lua_parameter_dictionary.h"
+#include "cartographer/common/port.h"
+#include "cartographer/mapping/map_builder.h"
+#include "cartographer_ros/sensor_bridge.h"
+
+namespace cartographer_ros {
+
+struct TrajectoryOptions {
```
Comment (yutakaoka) [inline cartographer_ros/cartographer_ros/trajectory_options.h]: I tried to use an encoded binary proto in ROS message.

File: cartographer_ros/cartographer_ros/map_options.h [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+#include <string>
+
+#include "cartographer/common/lua_parameter_dictionary.h"
+#include "cartographer/common/port.h"
+#include "cartographer/mapping/map_builder.h"
+#include "cartographer_ros/sensor_bridge.h"
+
+namespace cartographer_ros {
+
+struct MapOptions {
```
Comment (yutakaoka) [inline cartographer_ros/cartographer_ros/map_options.h]: Done

File: cartographer_ros/cartographer_ros/node_options.cc (hunk 1 above)
Comment (yutakaoka) [inline cartographer_ros/cartographer_ros/node_options.cc]: Done

File: cartographer_ros/cartographer_ros/node.h (hunk 6 above)
Comment (SirVer) [inline cartographer_ros/cartographer_ros/node.h]: Yes, a vector that bunches them up is preferable them. It conveys the semantics better.

**Comment by SirVer** [comment]
I will review carefully early next week.

File: cartographer_ros/cartographer_ros/node.h (hunk 6 above)
Comment (yutakaoka) [inline cartographer_ros/cartographer_ros/node.h]: Done

**Comment by SirVer** [comment]
Yutaka, thanks for your patience!

I reviewed the code and my remaining concerns are around malicious ROS service calls crashing the node. I left comments and would love if you could fix these in a followup PR.

I also changed some nits and added a few TODOs and comments here and there in https://github.com/googlecartographer/cartographer_ros/pull/346/commits/dbee73d1abb55f1934d2b0d5eaa84dd01fbf1c99 - please review this commit if you agree with all of them. In https://github.com/googlecartographer/cartographer_ros/pull/346/commits/a5bf65a7da047f63c64ebd59ed0ded5d92680b44 I just ran clang-format.

Let me know if you are happy with my changes - if you are I will merge.

Thanks for your contribution!

**Comment by yutakaoka** [comment]
Hi, thank you for fixing the code. Looks good to me.
I will run clang-format beforehand from next PR.

**Comment by SirVer** [comment]
I did not press send yesterday, sorry for that - please look through my remaining comments and see what you want to do there?

I am merging this one as is for now. Thanks for the contribution!

**Pull request merged**